pkg/kubelet/kubelet_pods.go

```diff
@@ -2435,6 +2436,24 @@ func (kl *Kubelet) convertToAPIContainerStatuses(pod *v1.Pod, podStatus *kubecon
 			if s != nil && s.State == kubecontainer.ContainerStateExited && s.ExitCode == 0 {
 				continue
 			}
+			// When the pod is static, init container status might be lost.
+			// We can assume that the init containers are already completed if the main containers are created.
+			// When the init container is a sidecar container, it should be waiting (and will be restarted).
+			isSidecar := container.RestartPolicy != nil && *container.RestartPolicy == v1.ContainerRestartPolicyAlways
+			if s == nil &&
+				kubetypes.IsStaticPod(pod) &&
+				kuberuntime.HasAnyRegularContainerCreated(pod, podStatus) &&
+				!isSidecar &&
+				statuses[container.Name].State.Waiting != nil {
+				statuses[container.Name].State = v1.ContainerState{
+					Terminated: &v1.ContainerStateTerminated{
+						Reason:   "Completed",
```

Comment:
Is the effect of this is that kubelet will restart the main container to run the Init container again?

Reply:
> ok, so current behavior is a wrong status, not the restart of a main container? just confirming

Correct, the current behavior is not restarting the init containers nor the pod.
This PR explains well https://github.com/kubernetes/kubernetes/pull/108583.
> Since static pod has no podstatus saved except in mirror pod, so while init container GC'd and kubelet restarts, static pod can only generate its pod status from current runtime, but after https://github.com/kubernetes/kubernetes/pull/96572 merged, init containers would not be re-created again, so we should reset init-container status for static pod correctly.